### DISCUSSION TITLE
Is temperature the best variable to monitor? Or is energy?

### DISCUSSION THREAD

**Discussion by Ergonomicmike** [sb53systems/G-Code-Flow-Temperature-Controller#9] [Ideas]
This is not an Issue. It's for "Discussion." Just brainstorming for a possible Version 2.0.

I think that you are making a good start here with your Temperature Controller. And it might be that nothing more is needed.

My comment is that using temperature as a metric for materials that undergo phase change (or other type of non-linear or non-monotonic change) might not be the best variable to monitor/control. (Although it is the easiest and might be good enough for what we are trying to achieve here.)

As a simple example, when water is heated to boiling, the temperature will remain at 100 degrees (C) as the liquid changes to a gas. 

So if one were to measure only temperature while boiling water - and try to control it - one might conclude that nothing is changing. Or that nothing needs to be changed when the temperature stays at 100 degrees. Whereas energy still needs to be added to the liquid to make it boil.

The point is that temperature does not tell the whole story.

I'm thinking that after your Controller is operating well using temperature, that perhaps "energy" (joules) is what should be controlled/calculated. Or maybe even "power." 

For example, faster flowing filament requires more energy to keep the extrusion at the phase/viscosity that we want it. 

It requires more energy because there is more filament being melted per unit time (power) and also because more energy is being carried away (from the extruder) by the (melted) filament at high flow rates.

So the PWM to the heater should increase just before high filament flow and continue high until just before the filament flow returns to nominal. But how much energy should be added?

Now, it might be in the above example that even if we increase the energy to the heater, the measured temperature of the filament (actually, of the heat block) might not change in the short term. Especially if the heat block has a lot of thermal inertial. 

And so I am thinking that calculated energy or power needed is a better metric for consistent flow/appearance.

Unfortunately I'm not smart enough to do the math. But I offer the idea for a Version 2.0 if desired.

**Comment by Nathan22211** [comment]
I don't think 3d printers melt plastic to a liquid state though. There are some that change structure when hot enough e.g. PET, HDPE.

**Comment by sb53systems** [comment]
Hey,
I'm not a specialist in polymer extrusion, but I think both solutions will do the same!  
the filament absorbs the energy gradually in the heating block until it comes out of the nozzle with a final temperature (printing PLA at 250/270° and Height speed, does not mean that it has reached this temperature), it is necessary to calibrate the temperatures used according to the appearance or behavior of the extruded filament, not according to the maximum flow.  
  
Thank you for your participation.  
Happy Sart 3D Printing.
Salim.

**Comment by Ergonomicmike** [comment]
Since I am brainstorming, another possible way to feed-forward this is to use Klipper's feature that provides the Temperature State (of the extruder), expressed as a percentage of power. (PWM I presume.) 

A user would take two measurements: First, the average PWM percentage at a set temperature (say 220 C for PLA), not printing anything, after steady state PWM is reached. Call this the "idle" PWM value. (PID tuning might affect how much this value perturbs around average.)

Second, then the user extrudes PLA at max flow rate, say 12 mm^3/sec for, say, 30 seconds. (Or until steady state PWM is reached.) The average PWM at the end of the test is the "max" PWM value. (A user could use a snippet from Stefan's code to perform this test.)

So then, for sections of a print that will print at the max flow rate, we would command the max PWM value for that section. (Probably with a lead-in and lead-out time to counter thermal inertia.) 

For a section with half the max flow rate, then we would command the PWM to be = idle PWM + ½ max PWM. (Here I'm assuming a linear relationship between extra energy needed (above idle) and flow rate. I don't expect that it's linear because real life doesn't scale linearly with volume. But a linear approximation might be good enough here for what we want to accomplish.) 

If one were to use PWM values, then perhaps Energy Compensation should happen in Klipper, as opposed to happening in slicers. (Because slicers can't control PWM. But Klipper can.)

**Comment by sb53systems** [comment]
@Ergonomicmike, I thank you for these interesting proposals, and encourage you to check the feasibility of the PMW change during printing, direct me if so.  
I'm not as good a programmer and I can't stay much time on this project, because it's supposed to be implemented in Slicers.  
This project is just supposed to take the first step and show the benefits.  
I remain at your disposal if you need more information.  

Best regards.
Salim.

**Comment by Ergonomicmike** [comment]
@sb53systems No worries. My hope is that someone from Klipper will see your work and my idea and, if possible, use them to improve 3d printing for all of us.

**Comment by Nathan22211** [comment]
Found this out, there is a max power param to set heater temperature.
Danger klipper can also change PID values, though it's MPC feature might
fare better with this script.

Also any chance of Linux support. I don't think Pascal is hard to compile
on Linux so I'm tempted to do it myself

**Comment by sb53systems** [comment]
Let me know if you have trouble compiling on other platforms.
Good luck.

**Comment by dbaarda** [comment]
I think the energy bit is a very important point.

I've done some work on temp control, but in my case it was not to optimize the fillament temp *when exiting* the nozzle, it was for optimizing the cooling *after exiting* the nozzle to prevent sag/drool and give good layer adhesion. I used a model of the heat-energy in the extruded fillament based on flow assuming it was at the nozzle set-temp and then adjusted fan speeds to try and achive the desired cooling rate for the cooling effects of the forced air flow on the printed surface area.

Optimizing the nozzle set-temp and flow-rates to try and maintain the ideal filament-exit-temp is not something I even considered bothering with for a few reasons;

1. The printer's PID controller already tries to ensure the fillament is at the set-temp on exiting the nozzle, adjusting the energy pumped into it to maintain the nozzle-temp at that temperature.
2. My printer is old and slow, so it's nowhere near approaching the speed limits of filament heating and flow.

In an ideal world, the ideal filament exit-temp would be constant, the measured nozzle-temp and exit-temp would be identical, and the PID control would perfectly maintain the nozzle-temp at the set-temp. In practice the ideal exit-temp might vary with flow-rate or line-profile (role/overhang/etc), there is a temperature gradient between the nozzle's temperature sensor and the exiting filament resulting in a small difference that will vary with the flow-rate, and the PID controller only reacts to temperature changes so changes to the energy pumped in lag behind changes to the flow-rate. PID controllers also include an integral term which gives them a "short-term-memory" of the past temps which can make them "overshoot" or "undershoot"  after a period of doing the opposite, so it's not as simple as just lag, but it's mostly lag.

In practice this means that the exit-temp stabilizes lower than the set-temp by an amount that varys with stabilized flow-rate, and it also dips when flow-rate increases and rises when flow-rate decreases. The magnitude of these differences would vary hugely with the printer nozzle geometry, heater power, and PID control settings. They also become more significant at higher flow-rates, so they probably only start to matter if your printer is fast enough. As for what the ideal exit-temp is based on flow-rate and line-profile, I have no idea and will leave that for others.

This script attempts to maintain the ideal exit-temp by adjusting the set-temp and flow-rate. I'm not exactly sure of the formula's it uses, but I believe it varys set-temp linearly between a range based on average flow-rate, and adjusts print-speeds to smooth out flow-rate changes. The set-temp changes should help compensate for the constant flow-rate difference, and smoothing flow-rate changes would help reduce the changing flow-rate differences. I think it's mostly been developed and tuned based on experimenting with results, which is fine, but might have missed things and probably will not generalize well to other printers. It's also really hard to isolate temp-related benefits from other types of benefits in experiments varying print-speeds.

I think it should be possible to design a simple thermal model for how set-temp, nozzle-temp, and exit-temp interact depending on flow-rate. This could then be used to adjust the set-temp and print-speeds to get the desired exit-temp. I think it will show that you can compensate pretty well for flow-rate changes with pre-emptive set-temp changes and get similar results without having to reduce speeds as much as you do with flow-rate smoothing. It might also show the best set-temp relationship with flow-rate is non-linear.

I might have a go at figuring out a model some time in the next week or so. I also have my own gcode-simulation framework for post-processing scripts written in Python which is a bit more linux-friendly than Delphi at https://github.com/dbaarda/workspace-tools/tree/main/3DPrinting (it's a bit of a mess and not ready for general use, sorry about that). I might play with implementing it in that.
